New keys for major mode: refresh and specialized quit-window

bmag opened the pull request:
- g: refresh imenu-list entries according to recently indexed buffer
- q: now also disables imenu-list-minor-mode

#26

colinxy commented:
`imenu-list-quit-window` should be an interactive function.

colinxy commented:
If I kill the `*Ilist*` buffer, when I reenable imenu-list, it won't show anything, and `g` does not work to refresh it.

colinxy commented:
I think a fix is on line 482. It is unnecessary to check whether entries have changed, since this emacs buffer does not write to disk, so I think just inserting them in buffer should be fast enough, and it solves the corner case that I have mentioned above (when I accidentally kill the buffer).

So removing the `unless` clause around it should solve the problem.

```emacs lisp
(unless (equal old-entries imenu-list--imenu-entries)
  (with-current-buffer (imenu-list-get-buffer-create)
    (imenu-list-insert-entries)))
```

bmag commented:
I've updated the code with a few fixes:

I've added the missing `interactive` declaration.

I added an optional `force-update` argument to `imenu-list-update`, and `imenu-list-refresh` now uses it.

I added a new command: `imenu-list-smart-toggle`, to replace the `imenu-list-minor-mode` command (from UI perspective). It enables/disables `imenu-list-minor-mode` depending on the visibility of the Ilist buffer. Can you try binding and using it instead of `imenu-list-minor-mode`, and tell me if there are any problems? It should fix the issue of calling `imenu-list-minor-mode` twice. (note that it's a new function and won't have an autoload on your machine, so you will need to `(require 'imenu-list)` in your config)

The `unless` clause is there because inserting the entries also moves point in Ilist buffer to the entry that matches the point location in the indexed buffer. Think of this use-case: I scroll (with the mouse or arrow keys) the Ilist buffer to view other entires. If we remove the `unless` clause, then point will jump back whenever I pause scrolling for a second.
Instead, I modified the condition to check if the buffer is alive. If it's not (e.g. the user accidentally  killed it), then the entries are re-inserted.

colinxy commented:
Just 1 more problem, `force-update` argument is not used at all in `imenu-list-update` function, so `g` still does not work properly.

Everything else is working fine, and `imenu-list-smart-toggle` is a great idea.

colinxy commented:
Works great, thanks for the feature!

bmag commented:
Thanks for testing the changes. I've re-organized the commits and pushed to master. New version should be available via MELPA in a few hours.
